### PR TITLE
Prevent MSVC 'needs to have a dll interface' warnings

### PR DESCRIPTION
**Pull request by swebb2066**
This commit is progress toward resolving [LOGCXX-558](https://issues.apache.org/jira/browse/LOGCXX-558)

### CODE REVIEW

File: src/main/include/log4cxx/log4cxx.h.in [hunk 1]
```diff
@@ -64,8 +64,15 @@ typedef unsigned int log4cxx_uint32_t;
 	typedef std::weak_ptr<T> T##WeakPtr
 #define LOG4CXX_UNIQUE_PTR_DEF(T) typedef std::unique_ptr<T> T##UniquePtr;
 #define LOG4CXX_LIST_DEF(N, T) typedef std::vector<T> N
+#define LOG4CXX_PRIVATE_PTR(T) std::unique_ptr<T>
```
Comment (rm5248) [inline src/main/include/log4cxx/log4cxx.h.in:67]: Is this define really needed?

File: src/main/include/log4cxx/log4cxx.h.in [hunk 2]
```diff
@@ -76,8 +83,11 @@ typedef unsigned int log4cxx_uint32_t;
 //    definitions used when using DLL
 #define LOG4CXX_EXPORT __declspec(dllimport)
 #endif
+#define LOG4CXX_INSTANTIATE_EXPORTED_PTR(T) template class LOG4CXX_EXPORT std::shared_ptr<T>
 #else
 #define LOG4CXX_EXPORT
+#define LOG4CXX_DECLARE_PRIVATE_MEMBER_PTR(T, V) struct T; LOG4CXX_PRIVATE_PTR(T) V;
+#define LOG4CXX_INSTANTIATE_EXPORTED_PTR(T)
```
Comment (rm5248) [inline src/main/include/log4cxx/log4cxx.h.in:90]: Do we need to have two versions of this macro?  I don't know enough about how windows works, but would it be possible to just add `LOG4CXX_EXPORT` to the already-existing `LOG4CXX_PTR_DEF` macro?

**Review by rm5248 — COMMENTED** [review]
Looks pretty good to me, I'm just a little confused as to the macro definitions.

Please squash your commits when merging.

File: src/main/include/log4cxx/log4cxx.h.in (hunk 1 above)
Comment (swebb2066) [reply src/main/include/log4cxx/log4cxx.h.in:67]: I only wanted the definitions to be in the same place, so the usage is (for example)
		LOG4CXX_DECLARE_PRIVATE_MEMBER_PTR(AppenderSkeletonPrivate, m_priv)
		AppenderSkeleton(LOG4CXX_PRIVATE_PTR(AppenderSkeletonPrivate) priv);

File: src/main/include/log4cxx/log4cxx.h.in (hunk 2 above)
Comment (swebb2066) [reply src/main/include/log4cxx/log4cxx.h.in:90]: gcc says instantiated templates must be in their own namespace (not sure how to interpret that when the template and the template parameter are in different namespaces). Visual Studio does not issue that error, but not instantiating the template at all results in a  'needs to have a dll interface' warning.

My solution was to only instantiate when using Visual Studio.

See for example, in propertysetter.h

File: src/main/include/log4cxx/log4cxx.h.in (hunk 1 above)
Comment (rm5248) [reply src/main/include/log4cxx/log4cxx.h.in:67]: It feels like it doesn't really add anything in terms of making things simpler to read(to me).

Not really a huge issue, more of a preference thing in how to organize the code.

**Pull request merged**